python/ray/data/tests/test_strict_in_virtual_cluster.py

```diff
@@ -0,0 +1,362 @@
+    )
+    for actor_id, actor_info in job_actors.items():
+        assert actor_info["address"]["rayletId"] in target_virtual_cluster_node_ids
+
+
+@pytest.mark.parametrize(
+    "create_virtual_cluster",
+    [
+        {
+            "node_instances": [("1c2g", 2), ("2c4g", 2), ("8c16g", 4)],
+            "virtual_cluster": {
+                "VIRTUAL_CLUSTER_0": {
+                    "allocation_mode": "mixed",
+                    "replica_sets": {
+                        "1c2g": 2,
+                    },
+                },
+                "VIRTUAL_CLUSTER_1": {
+                    "allocation_mode": "mixed",
+                    "replica_sets": {
+                        "2c4g": 2,
+                    },
+                },
+                "VIRTUAL_CLUSTER_2": {
+                    "allocation_mode": "mixed",
+                    "replica_sets": {
+                        "8c16g": 4,
+                    },
+                },
+            },
+        }
+    ],
+    indirect=True,
+)
+def test_auto_parallelism(create_virtual_cluster):
+    _, job_client = create_virtual_cluster
+    MiB = 1024 * 1024
+    TEST_CASES = [
+        (1024, (4, 8, 64)),  # avail_cpus * 2
+        (10 * MiB, (10, 10, 64)),  # MAX_PARALLELISM, MAX_PARALLELISM, avail_cpus * 2
+    ]
+
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        for i in range(3):  # 3 virtual clusters in total
+            signal_actor_name = f"storage_actor_{i}"
+            signal_actor = JobSignalActor.options(
+                name=signal_actor_name, namespace="storage", num_cpus=0
+            ).remote()
+            ray.get(signal_actor.is_ready.remote())
+            for data_size, expected_parallelism in TEST_CASES:
+                driver_script = """
+import ray
+from ray.data._internal.util import _autodetect_parallelism
+from ray.data.context import DataContext
+
+
+data_size = {data_size}
+signal_actor_name = "{signal_actor_name}"
+print("Job is running, data_size: ", data_size, "signal_actor_name: ", signal_actor_name)
+ray.init(address="auto")
+signal_actor = ray.get_actor(signal_actor_name, namespace="storage")
+ray.get(signal_actor.ready.remote())
+target_max_block_size = DataContext.get_current().target_max_block_size
+class MockReader:
+    def estimate_inmemory_data_size(self):
+        return data_size
+
+
+final_parallelism, _, _ = _autodetect_parallelism(
+    parallelism=-1,
+    target_max_block_size=target_max_block_size,
+    ctx=DataContext.get_current(),
+    datasource_or_legacy_reader=MockReader(),
+)
+
+ray.get(signal_actor.data.remote(final_parallelism))
+ray.get(signal_actor.unready.remote())
+"""
+                driver_script = driver_script.format(
+                    signal_actor_name=signal_actor_name, data_size=data_size
+                )
+                submit_job_to_virtual_cluster(
+                    job_client, tmp_dir, driver_script, f"VIRTUAL_CLUSTER_{i}"
+                )
+                # wait for job running
```

Comment:
add a notify here to make sure the following code is executed before the driver code